### ISSUE TITLE
Qt 6 name conflict

### ISSUE THREAD

**Issue by flaviotordini** [simonbrunel/qtpromise#37]
Hello, I recently discovered this project and I'm amazed by it. I'm planning to start using QPromise in my own project. But Qt 6 is coming and has a new class named QPromise. Do you plan to change the class name? Do you plan to support Qt 6 at all?

**Comment by simonbrunel** [comment]
Thanks @flaviotordini!

> Do you plan to change the class name?

I'm not sure there is any name conflict since the `QPromise` class of this library is under the `QtPromise` namespace, while the Qt one is under the `QtCore` namespace. So as long as you don't `use namespace QtPromise;` it should be fine, right? Also, headers are different, `#include <QtPromise>` vs `#include <QPromise>`.

I didn't try Qt6 yet, so maybe I missed something. But if it works by using the `QtPromise` namespace explicitly, then there is no reason to rename this class. You should also be able to use both class in the same file:

```cpp
#include <QtPromise>
#include <QPromise>

//...

QtPromise::QPromise p; // This library promise
QPromise; // The Qt promise
```

> Do you plan to support Qt 6 at all?

I would be happy to port that lib to Qt6 but does it make sense now that Qt provides an official promise implementation? if yes, why?

**Comment by flaviotordini** [comment]
I really don't know. I was just wondering whether using this library now makes sense now with Qt 6 coming. I see that QPromise in Qt 6 is about multithreading which I am not interested in. I also don't like their QFuture/QFutureWatcher stuff. Too complex (because of multithreading). I would like my promises to not spawn or use different threads so I thought this library would also be useful in the Qt 6 future.

**Comment by pwuertz** [comment]
We did participate in some sort of discussion initiated by the Qt developers concerning the inclusion of a promise standard in Qt6. The discussion thread went quiet at some point, so I'm assuming that there are no plans to adopt a promise workflow in the first few Qt 6.X releases.

Maybe we will see some kind of movement in the Qt base once C++20 / co-routine support is more established. I did some experiments and at least for linear continuations C++20 looks pretty good already. But even with co-routines as a basis you'd still need a library like QtPromise for handling all the N-to-N continuation situations you encounter from time to time.

**Comment by j-ulrich** [comment]
> I see that QPromise in Qt 6 is about multithreading [...]

Are you sure? To me it looks like `QPromise` could also be used without threads. I mean, you can simply call the `QPromise()` constructor and return the `QPromise::future()` to callers. And using a `QFutureWatcher` you can "wait" for the result using the event loop (signals and slots).

**Comment by pwuertz** [comment]
Oh you're right. Qt6 Core indeed now provides continuations (sync/async/deferred/threaded) via [QFuture::then](https://doc-snapshots.qt.io/qt6-dev/qfuture.html#then-1). Semantics are just a bit different from ECMAScript / QtPromise. `QtCore/QFuture` is the promise-like object, `QtCore/QPromise` is the resolver/rejecter. Also, in Qt6 the resolver is created first and the promise is obtained from it (I prefer this pattern over the JS style tbh).

**Comment by simonbrunel** [comment]
So, is there any benefits to port that lib to Qt6 instead of using the new `QFuture/QPromise` APIs?

> Also, in Qt6 the resolver is created first and the promise is obtained from it (I prefer this pattern over the JS style tbh).

It's/was also possible in JavaScript, using the [obsolete deferred approach](https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Deferred). I worked with both methods and I personally prefer the constructor style since I found it easier to maintain and less error prone.

**Comment by dpurgin** [comment]
@simonbrunel Hey Simon, if you missed it, here's the blog posting about Qt 6 promises: https://www.qt.io/blog/asynchronous-apis-in-qt-6

I personally think that we have at least 3 to 5 years until Qt 6 settles in and it is polished enough to be seriously used as the "default" Qt version to start new projects with. Until that, I'll definitely stick with Simon's QtPromise.

**Comment by simonbrunel** [comment]
Thanks @dpurgin! I'm not sure I will have enough time to port the lib to Qt6 but I will be happy to review PR if you guys want to start migrating it. It would also be the opportunity to introduce breaking changes if it allows a better integration with the new Qt API.

**Comment by pwuertz** [comment]
I can report that QtPromise works perfectly fine in Qt6. All you have to do is to watch out for name-clashes. E.g. I often imported `QtPromise::QPromise` into my global namespace and had to replace it by `using` an alias that isn't clashing with global `QPromise` from Qt6, or by using the fully qualified name with the namespace prefix.

**Comment by pwuertz** [comment]
For building or including QtPromise with CMake, a variable for selecting the Qt package to be imported comes in handy. Something like this perhaps: https://github.com/pwuertz/qtpromise/commit/160a4b8de370da7f0e94860732b893f900d8baf8

**Comment by simonbrunel** [comment]
I pushed [changes](https://github.com/simonbrunel/qtpromise/commit/9f01c130cd061d151af80fa4d7ee7fc6f1f18ba0) to make this project compatible with Qt6. It requires just a few minor fixes but the main issue was indeed the name conflict with `QPromise` from `QtCore`, thus I had to get rid of all the `using namespace QtPromise`.

@pwuertz  I used the same approach as the one I found in [this fork](https://github.com/marc-cpdesign/qtpromise/commit/a61a77610f6c03762270742fb98fcc75d53c522f) (thanks @marc-cpdesign).
